shell/platform/windows/keyboard_key_handler.h

```diff
@@ -0,0 +1,173 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_WINDOWS_KEYBOARD_KEY_HANDLER_H_
+#define FLUTTER_SHELL_PLATFORM_WINDOWS_KEYBOARD_KEY_HANDLER_H_
+
+#include <deque>
+#include <memory>
+#include <string>
+
+#include "flutter/shell/platform/common/client_wrapper/include/flutter/basic_message_channel.h"
+#include "flutter/shell/platform/common/client_wrapper/include/flutter/binary_messenger.h"
+#include "flutter/shell/platform/windows/keyboard_handler_base.h"
+#include "flutter/shell/platform/windows/public/flutter_windows.h"
+#include "rapidjson/document.h"
+
+namespace flutter {
+
+class FlutterWindowsView;
+
+// Handles key events.
+//
+// This class detects whether an incoming event is a redispatched one,
+// dispatches native events to delegates and collect their responses,
+// and redispatches events unhandled by Flutter back to the system.
+// See |KeyboardHook| for more information about dispatching.
+//
+// The exact behavior to handle events are further forwarded into
+// delegates. See |KeyboardKeyHandlerDelegate| and its subclasses.
+class KeyboardKeyHandler : public KeyboardHandlerBase {
+ public:
+  // An interface for concrete definition of how to asynchronously handle key
+  // events.
+  class KeyboardKeyHandlerDelegate {
+   public:
+    // Defines how to how to asynchronously handle key events.
+    //
+    // |KeyboardHook| should invoke |callback| with the response (whether the
+    // event is handled) later for exactly once.
```

Comment:
> Or are these a different kind of delegate? 

There are 2 layers:
- A `FlutterWindowsView` owns multiple `KeyboardHandlerBase`s.
  - Such as `KeyboardKeyHandler` and `TextInputPlugin`. 
  - Events are propagated along the handler list synchronously in insertion order.
  - Each handler decides whether to pass the event to the next handler synchronously.
- A `KeyboardKeyHandler` owns multiple `KeyboardKeyHandlerDelegate`s.
  - Such as `KeyboardKeyEmbedderHandler` and `KeyboardKeyChannelHandler`.
  - Events are propagated to all delegates immediately in insertion order.
  - The `handled` result is collected asynchronously, combined to decide the overall result for `KeyboardKeyHandler`.

> How is the order of handling determined?

Insertion order. (I tried to add a unittest for `FlutterWindowsView` but it basically requires mocking `FlutterWindowsEngine` and even adding some pseudo implementations. I don't know if it's worth doing so just to preserve the order between 2 lines.

> Can one handler prevent the next from looking at it? 

No you can't. Basically all events are sent through the embedder API, then sent through the channel API, unconditionally. It's designed this way so that legacy API continues to work.

Reply:
But what happens if, on the framework side, people are using both the old and the new way of doing things? Won't that cause double-handling of those events?

Maybe it doesn't matter that much because once the framework focus system is using the new way, only people explicitly not using the focus system will be affected this way, and we expect those to be few and far between?

Reply:
I think the new way wouldn't prevent anything the old way can do. The only "prevent propagation" that comes with the framework is the focus system, which has been migrated. The only reason there are two systems running at the same time is for backward compatibility.